### ISSUE TITLE
etcd services pods error out due to PodInitialization on 4.8 ocp cluster

### ISSUE THREAD

**Issue by prietyc123** [openshift/odo#4728]
/kind bug

## What versions of software are you using?

**Operating System:**
All supported

**Output of `odo version`:**
main

## How did you run odo exactly?

```
[odo] time="2021-05-18T05:16:48Z" level=info msg="Don't start the stopped program because its retry times 0 is greater than start retries 0" program=devrun 
Running oc with args [oc get pods -n cmd-service-test218byx]
[oc] NAME                                     READY   STATUS     RESTARTS   AGE
[oc] example-jgg6bc6ppk                       0/1     Init:0/1   0          57s
[oc] nodejs-x932449387-lpai-5f68c4786-mr85p   1/1     Running    0          59s
Running oc with args [oc get pods -n cmd-service-test218byx]
[oc] NAME                                     READY   STATUS     RESTARTS   AGE
[oc] example-jgg6bc6ppk                       0/1     Init:0/1   0          59s
[oc] nodejs-x932449387-lpai-5f68c4786-mr85p   1/1     Running    0          61s
Running oc with args [oc get pods example-jgg6bc6ppk -o template="{{.status.phase}}" -n cmd-service-test218byx]
[oc] "Pending"Running oc with args [oc get pods example-jgg6bc6ppk -o template="{{.status.phase}}" -n cmd-service-test218byx]
```

## Actual behavior

etcd services pods are not initialised.

## Any logs, error output, etc?
details: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-odo-main-v4.8-integration-e2e-periodic/1394442840263626752#1:build-log.txt%3A1117

**Comment by prietyc123** [comment]
I tried debugging it locally but didn't find valid reason of the failure.

My observations so far:

```
# All pods in the namespace "cmd-service-test223kyo", last executed tests
$ oc get pods -n cmd-service-test223kyo
NAME                                          READY   STATUS     RESTARTS   AGE
pod/example-td6dkzgdqk                        0/1     Init:0/1   0          17m
pod/nodejs-x402983414-kvkw-84d664477f-rbkdl   1/1     Running    0          17m

# Describe pod which is not getting initialised
$ oc describe pod example-td6dkzgdqk -n cmd-service-test223kyo
Name:         example-td6dkzgdqk
Namespace:    cmd-service-test223kyo
Labels:       app=etcd
              etcd_cluster=example
              etcd_node=example-td6dkzgdqk
Annotations:  etcd.version: 3.2.13
              [...]
Status:       Pending
[...]
Controlled By:  EtcdCluster/example
Init Containers:
  check-dns:
    Container ID:  cri-o://39e52b3b1546d466e658e4034d6e226441ab874c1d6f5a0594cd6358139b0026
    Image:         busybox:1.28.0-glibc
    Image ID:      docker.io/library/busybox@sha256:0b55a30394294ab23b9afd58fab94e61a923f5834fba7ddbae7f8e0c11ba85e6
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -c
      
                TIMEOUT_READY=0
                while ( ! nslookup example-td6dkzgdqk.example.cmd-service-test223kyo.svc )
                do
                  # If TIMEOUT_READY is 0 we should never time out and exit 
                  TIMEOUT_READY=$(( TIMEOUT_READY-1 ))
                              if [ $TIMEOUT_READY -eq 0 ];
                                  then
                                      echo "Timed out waiting for DNS entry"
                                      exit 1
                                  fi
                              sleep 1
                            done
    State:          Running
      Started:      Tue, 18 May 2021 12:06:00 +0530
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:         <none>
Containers:
  etcd:
    Container ID:  
    Image:         quay.io/coreos/etcd:v3.2.13
    Image ID:      
   [...]
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Liveness:       exec [/bin/sh -ec ETCDCTL_API=3 etcdctl endpoint status] delay=10s timeout=10s period=60s #success=1 #failure=3
    Readiness:      exec [/bin/sh -ec ETCDCTL_API=3 etcdctl endpoint status] delay=1s timeout=5s period=5s #success=1 #failure=3
    Environment:    <none>
    Mounts:
      /var/etcd from etcd-data (rw)
Conditions:
  Type              Status
  Initialized       False 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  etcd-data:
    Type:        EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:      
    SizeLimit:   <unset>
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason          Age        From                                                Message
  ----    ------          ----       ----                                                -------
  Normal  Scheduled       <unknown>                                                      Successfully assigned cmd-service-test223kyo/example-td6dkzgdqk to <node>
  Normal  AddedInterface  6m10s      multus                                              Add eth0 [<ip>/<port>]
  Normal  Pulled          6m9s       kubelet, <node>  Container image "busybox:1.28.0-glibc" already present on machine
  Normal  Created         6m9s       kubelet, <node>  Created container check-dns
  Normal  Started         6m9s       kubelet, <node>  Started container check-dns

# Events to that pod
$ oc get events -n cmd-service-test223kyo
LAST SEEN   TYPE     REASON              OBJECT                                         MESSAGE
17m         Normal   CreatedSCCRanges    namespace/cmd-service-test223kyo               created SCC ranges
16m         Normal   New Member Added    etcdcluster/example                            New member example-td6dkzgdqk added to cluster
16m         Normal   Scheduled           pod/example-td6dkzgdqk                         Successfully assigned cmd-service-test223kyo/example-td6dkzgdqk to <node>
16m         Normal   AddedInterface      pod/example-td6dkzgdqk                         Add eth0 [<ip>/<port>]
16m         Normal   Pulled              pod/example-td6dkzgdqk                         Container image "busybox:1.28.0-glibc" already present on machine
16m         Normal   Created             pod/example-td6dkzgdqk                         Created container check-dns
16m         Normal   Started             pod/example-td6dkzgdqk                         Started container check-dns
16m         Normal   Scheduled           pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Successfully assigned cmd-service-test223kyo/nodejs-x402983414-kvkw-84d664477f-rbkdl to <node>
16m         Normal   AddedInterface      pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Add eth0 [<ip>/<port>]
16m         Normal   Pulled              pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Container image "registry.access.redhat.com/ocp-tools-4/odo-init-container-rhel8:1.1.10" already present on machine
16m         Normal   Created             pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Created container copy-supervisord
16m         Normal   Started             pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Started container copy-supervisord
16m         Normal   Pulling             pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Pulling image "registry.access.redhat.com/ubi8/nodejs-14:latest"
16m         Normal   Pulled              pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Successfully pulled image "registry.access.redhat.com/ubi8/nodejs-14:latest" in 530.984654ms
16m         Normal   Created             pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Created container runtime
16m         Normal   Started             pod/nodejs-x402983414-kvkw-84d664477f-rbkdl    Started container runtime
16m         Normal   SuccessfulCreate    replicaset/nodejs-x402983414-kvkw-84d664477f   Created pod: nodejs-x402983414-kvkw-84d664477f-rbkdl
16m         Normal   ScalingReplicaSet   deployment/nodejs-x402983414-kvkw              Scaled up replica set nodejs-x402983414-kvkw-84d664477f to 1

# Get logs of the pod
$ oc logs example-td6dkzgdqk -n cmd-service-test223kyo
Error from server (BadRequest): container "etcd" in pod "example-td6dkzgdqk" is waiting to start: PodInitializing
```

**Comment by prietyc123** [comment]
Initcontainer stuck in running state and not getting excited

```
$ oc logs example-td6dkzgdqk -c check-dns -n cmd-service-test223kyo
Server:    <server-ip>
Address 1: <server-ip> dns-default.openshift-dns.svc.cluster.local
nslookup: can't resolve 'example-td6dkzgdqk.example.cmd-service-test223kyo.svc'
```

**Comment by valaparthvi** [comment]
tldr; It seems like a problem with the init container image i.e. busybox:1.28.0-glibc, which goes on a CrashLoopBackoff.

Steps that were taken to reproduce this - 
1. Get a OCP4 cluster and install etcd operator in it.
2. Create a component - `odo create nodejs --starter`
3. Create a service for the component - `odo service create etcdoperator.v0.9.4-clusterwide/EtcdCluster`
4. Run `kubectl get pods` and check the `example-XXX` pod's status. It is in Init:0/1 state.
```
^C
➜  sboc kubectl get pods
NAME                                READY   STATUS             RESTARTS   AGE
busybox                             0/1     CrashLoopBackOff   11         34m
example-2qqgdbv2t8                  0/1     Init:0/1           0          36m
nodejs-sboc-fkhh-6c9749d54-klj5h    1/1     Running            0          36m
```

There was not much information in the pod `describe` except that the error is caused due to the Initcontainer. So I decided to run the init container image - `kubectl run busybox --image=busybox:1.28.0-glibc` which also resulted in error. 

So I ran `podman run busybox --image=busybox:1.28.0-glibc` which errored out with - 
```Error: container_linux.go:370: starting container process caused: exec: "--image=busybox:1.28.0-glibc": executable file not found in $PATH: OCI not found```

I assume the problem is with busybox image and not odo.

**Comment by dharmit** [comment]
How about trying to spin up the EtcdCluster instance using only it's yaml and kubectl?

You can get the YAML by doing `odo service create etcdoperator.v0.9.4-clusterwide/EtcdCluster --dry-run`. Use the output to start EtcdCluster using kubectl. Do you still see the same behaviour?

What about OCP 4.7 or minikube? Are you seeing similar behaviour on those clusters as well? Are the container images being pulled (busybox image that you've mentioned) the same across all the clusters?

I'm asking these questions to understand if the behaviour is same across all clusters. That way, it could indeed be a problem outside odo.

But then why are things not failing currently on OCP 4.7 which we use in our integration tests? It should fail there as well, no?

**Comment by valaparthvi** [comment]
It seems like the problem is not with the InitContainer Image, but with DNS resolution.
It is possible to lookup with the FQDN, in this case - <example-pod-name>.example.myproject.svc.cluster.local but not without the cluster domain, so <example-pod-name>.example.myproject.svc` does not seem to resolve.

This same thing works on OCP4.7, but I don't know how to replicate this on K8s yet.

**Comment by prietyc123** [comment]
> This same thing works on OCP4.7, but I don't know how to replicate this on K8s yet.
> 
> ```
> ➜  compocp4 kubectl exec -it example-lzjnr24xvc -c check-dns -- nslookup example-lzjnr24xvc.example.myproject.svc
> Server:    172.30.0.10
> Address 1: 172.30.0.10 dns-default.openshift-dns.svc.cluster.local
> 
> nslookup: can't resolve 'example-lzjnr24xvc.example.myproject.svc'
> command terminated with exit code 1
> ➜  compocp4 kubectl exec -it example-lzjnr24xvc -c check-dns -- nslookup example-lzjnr24xvc.example.myproject.svc.cluster.local
> Server:    172.30.0.10
> Address 1: 172.30.0.10 dns-default.openshift-dns.svc.cluster.local
> 
> Name:      example-lzjnr24xvc.example.myproject.svc.cluster.local
> Address 1: 10.128.2.12 example-lzjnr24xvc.example.myproject.svc.cluster.local
> ➜  compocp4 
> ```

@valaparthvi I suspect if there is configuration change on 4.8 cluster w.r.t dns configuration. Could you please try to verify on that area. Because its really strange that its working with suffix `cluster.local` but not simply.

**Comment by dharmit** [comment]
4.8 release is on July 12, 2021. Ideally, fix the issue before that.

**Comment by valaparthvi** [comment]
I tested this thing on ocp 4.7 and k8s(minikube) both of which ran successfully, so maybe it is an issue with ocp4.8.
So, if we are going to move away from etcd operator, doesn't it make to sense to close this issue because other operators, redis for e.g. ran successfully on ocp4.8.

**Comment by dharmit** [comment]
/close

Closing this in favour of our plan to move to redis operator in test suite. Re-open if it becomes relevant later.
